### PR TITLE
vLLM support

### PR DESCRIPTION
**Pull request by cpacker**
**Please describe the purpose of this pull request.**
Is it to add a new feature? Is it to fix a bug?

**How to test**
How can we test your PR during review? What commands should we run? What outcomes should we expect?

**Have you tested this PR?**
Have you tested the latest commit on the PR? If so please provide outputs from your tests.

**Related issues or PRs**
Please link any related GitHub [issues](https://github.com/cpacker/MemGPT/issues) or [PRs](https://github.com/cpacker/MemGPT/pulls).

**Is your PR over 500 lines of code?**
If so, please break up your PR into multiple smaller PRs so that we can review them quickly, or provide justification for its length.

**Additional context**
Add any other context or screenshots about the PR here.

### CODE REVIEW

File: memgpt/local_llm/vllm/api.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    prompt_tokens = count_tokens(prompt)
+    if prompt_tokens > context_window:
+        raise Exception(f"Request exceeds maximum context length ({prompt_tokens} > {context_window} tokens)")
+
+    # Settings for the generation, includes the prompt + stop tokens, max length, etc
+    request = settings
+    request["prompt"] = prompt
+    request["max_tokens"] = int(context_window - prompt_tokens)
+    request["stream"] = False
+    # TODO pass model?
```
Comment (cpacker) [inline memgpt/local_llm/vllm/api.py:22]: Not sure if the POST endpoint allows model as a parameter? Didn't see it in the documentation so probably requires reading the web server code

File: memgpt/local_llm/vllm/api.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    request["model"] = model
+
+    # Set grammar
+    if grammar is not None:
+        raise NotImplementedError
```
Comment (cpacker) [inline memgpt/local_llm/vllm/api.py:27]: Probably is supported, just need to make sure we pass it the correct way

File: memgpt/local_llm/vllm/api.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def get_vllm_completion(endpoint, model, prompt, context_window, settings={}, grammar=None):
+    """https://github.com/vllm-project/vllm/blob/main/examples/api_client.py"""
+    prompt_tokens = count_tokens(prompt)
+    if prompt_tokens > context_window:
+        raise Exception(f"Request exceeds maximum context length ({prompt_tokens} > {context_window} tokens)")
+
+    # Settings for the generation, includes the prompt + stop tokens, max length, etc
+    request = settings
```
Comment (cpacker) [inline memgpt/local_llm/vllm/api.py:18]: Do we also need to pass context length? Or is that done when the server is started (on the server-side)?

File: memgpt/local_llm/vllm/api.py (hunk 1 above)
Comment (fragro) [reply memgpt/local_llm/vllm/api.py:22]: @cpacker The model is loaded when you start the API server with the `--model` flag

File: memgpt/local_llm/vllm/api.py (hunk 3 above)
Comment (fragro) [reply memgpt/local_llm/vllm/api.py:18]: @cpacker The context length is set when you start the server so you do not need to pass it as an argument

**Review by sarahwooders — APPROVED** [review]
This works for me locally, so I think its good to merge.

**Pull request merged**